controllers/postgresql_controller.go

```diff
@@ -94,3 +100,54 @@ func (a *postgreSQLAdapter) getServiceType() string {
 func (a *postgreSQLAdapter) getDiskSpace() string {
 	return a.Spec.DiskSpace
 }
+
+func (a *postgreSQLAdapter) performUpgradeTaskIfNeeded(ctx context.Context, avn *aiven.Client, avnGen avngen.Client, old *service.ServiceGetOut) error {
+	var currentVersion string = old.UserConfig["pg_version"].(string)
+	var targetVersion string = *a.getUserConfig().(pguserconfig.PgUserConfig).PgVersion
+
+	// No need to upgrade if pg_version hasn't changed
+	if targetVersion == currentVersion {
+		return nil
+	}
+
+	task, err := avnGen.ServiceTaskCreate(ctx, a.getServiceCommonSpec().Project, a.getObjectMeta().Name, &service.ServiceTaskCreateIn{
+		TargetVersion: service.TargetVersionType(targetVersion),
+		TaskType:      service.TaskTypeUpgradeCheck,
+	})
+	if err != nil {
+		return fmt.Errorf("cannot create PG upgrade check task: %q", err)
+	}
+
+	didTaskSucceed, err := waitForTaskToComplete(ctx, func() (bool, *bool, error) {
+		t, getErr := avn.ServiceTask.Get(ctx, a.getServiceCommonSpec().Project, a.getObjectMeta().Name, task.TaskId)
+		if getErr != nil {
+			return true, nil, fmt.Errorf("error fetching service task %s: %q", t.Task.Id, getErr)
+		}
+
+		if t.Task.Success == nil {
+			return false, nil, nil
+		}
+		return true, t.Task.Success, nil
```

Comment:
Much better! Could you please return `true, *t.Task.Success, nil`.
The `nil` thing must end here, or it will make all code around check for nil to not to panic.